d3a_api_client/redis_market.py

```diff
@@ -37,16 +83,79 @@ def _command_received(msg):
                         for command_name in ["market_stats", "grid_fees", "dso_market_stats"]}
         channel_subs.update({
-            f'{self.area_id}/market-events/market': self._on_market_cycle,
-            f'{self.area_id}/events/finish': self._on_finish
+            f'{self._channel_prefix}/response/register_participant': self._on_register,
+            f'{self._channel_prefix}/response/unregister_participant': self._on_unregister,
+            f'{self._channel_prefix}/market-events/market': self._on_market_cycle,
+            f'{self._channel_prefix}/events/finish': self._on_finish
         })
+
         self.pubsub.subscribe(**channel_subs)
         self.pubsub.run_in_thread(daemon=True)
 
+    def _check_transaction_id_cached_out(self, transaction_id):
+        return transaction_id in self._transaction_id_buffer
+
+    def _on_register(self, msg):
+        message = json.loads(msg["data"])
+        self._check_buffer_message_matching_command_and_id(message)
+        self.area_uuid = message["device_uuid"]
+        self.is_active = True
+
+    def _on_unregister(self, msg):
+        message = json.loads(msg["data"])
+        self._check_buffer_message_matching_command_and_id(message)
+        self.is_active = False
+
+    def _check_buffer_message_matching_command_and_id(self, message):
+        if "transaction_id" in message and message["transaction_id"] is not None:
+            transaction_id = message["transaction_id"]
+            if not any(command in ["register", "unregister"] and "transaction_id" in data and
+                       data["transaction_id"] == transaction_id
+                       for command, data in self._blocking_command_responses.items()):
+                raise RedisAPIException("There is no matching command response in _blocking_command_responses.")
+        else:
+            raise RedisAPIException(
+                "the answer message does not contain a valid 'transaction_id' member.")
+
+    def select_aggregator(self, aggregator_uuid, is_blocking=True, unsubscribe_from_device_events=True):
+        if not self.area_uuid:
+            assert False
+        logging.info(f"Market: {self.area_slug} is trying to select aggregator {aggregator_uuid}")
+
+        transaction_id = str(uuid.uuid4())
+        data = {"aggregator_uuid": aggregator_uuid,
+                "device_uuid": self.area_uuid,
+                "type": "SELECT",
+                "transaction_id": transaction_id}
+        self.redis_db.publish(f'aggregator', json.dumps(data))
+        self._transaction_id_buffer.append(transaction_id)
+
+        if is_blocking:
+            try:
+                wait_until_timeout_blocking(
+                    lambda: self._check_transaction_id_cached_out(transaction_id)
+                )
+                logging.warning(f"Market: {self.area_slug} has selected "
+                                f"AGGREGATOR: {aggregator_uuid}")
```

Comment:
Minor but any specific reason to have "Aggregator" in  caps lock?

Reply:
You can actually ask yourself, you implemented this on 29.06.2020 :-)  (again, copy/paste from the device)
I dont mind it being Caps lock.

Reply:
> You can actually ask yourself, you implemented this on 29.06.2020 :-)

... and approved by you ;-), Nevertheless my point is rather than exactly copy/pasting if there is a potential improvement one shouldn't be shy to do so, looks good to have same pattern being followed throughout the message like here `Market ` and `AGGREGATOR ` .
Thanks!

Reply:
OK, changing "Market" to capital letters.